File: quickshop/app/src/main/java/com/avas/quickshop/flutter_utils/FlutterEvents.kt
package com.avas.quickshop.flutter_utils

/// Created by Aabhash Shakya on 19/11/2025
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.asSharedFlow
import kotlinx.serialization.Serializable
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class FlutterEventsRepository @Inject constructor() {

    private val _events = MutableSharedFlow<FlutterEvent>()
    val events = _events.asSharedFlow()

    suspend fun emit(event: FlutterEvent) {
        _events.emit(event)
    }
}

//types of events coming from Flutter
sealed class FlutterEvent {
    data class CacheUUID(val uuid: String) : FlutterEvent()

    data class ProductSelected(val product: Product) : FlutterEvent()
    object UserLoggedOut : FlutterEvent()
}


@Serializable
data class Product(
    val id: Int,
    val title: String,
    val price: Double,
    val description: String,
    val category: String,
    val image: String,
    val rating: Rating? = null
)

@Serializable
data class Rating(
    val rate: Double,
    val count: Int
)

File: QuickShop-Ios/QuickShop-Ios/flutter_utils/FlutterEvents.swift
//
//  FlutterEvents.swift
//  quickshop-ios
//
//  Created by Aabhash Shakya on 22/11/2025.
//

import Foundation
import Combine

@MainActor
final class FlutterEventsRepository: ObservableObject {
    static let shared = FlutterEventsRepository()

    @Published var latestEvent: FlutterEvent? = nil

    func emit(_ event: FlutterEvent) {
        self.latestEvent = event
    }
}

enum FlutterEvent {
    case productSelected(Product)
    case cacheUUID(String)
    case userLoggedOut
}


//product model
struct Rating: Codable {
    let rate: Double?
    let count: Int?
}

struct Product: Codable, Identifiable {
    let id: Int
    let title: String
    let price: Double
    let description: String
    let category: String
    let image: String
    let rating: Rating?

    static func from(jsonString: String) -> Product? {
        guard let data = jsonString.data(using: .utf8) else { return nil }
        return try? JSONDecoder().decode(Product.self, from: data)
    }
}
